synapse/lib/ast.py

```diff
@@ -1907,186 +2033,224 @@ class FormPivot(PivotOper):
+    def buildgenr(self, runt, name):
 
-                refsprop = destform.props.get(refsname)
-                async for pivo in runt.snap.nodesByPropValu(refsprop.full, '=', node.ndef):
-                    yield pivo, path.fork(pivo)
+        if isinstance(name, list) or (prop := runt.model.props.get(name)) is None:
+
+            proplist = None
+            if isinstance(name, list):
+                proplist = name
+
+            if proplist is None:
+                proplist = runt.model.formsbyiface.get(name)
+
+            if proplist is None:
+                proplist = runt.model.ifaceprops.get(name)
+
+            if proplist is None and name.endswith('*'):
+                pref = name[:-1]
+                proplist = []
+                for form in runt.model.forms:
+                    if form.startswith(pref):
+                        proplist.append(form)
+
+                if not proplist:
+                    mesg = f'No forms match pattern {name}.'
+                    raise self.kids[0].addExcInfo(s_exc.NoSuchForm(mesg=mesg, name=name))
+
+            if proplist is None:
+                mesg = f'No property named {name}.'
+                raise self.kids[0].addExcInfo(s_exc.NoSuchProp(mesg=mesg, name=name))
+
+            pgenrs = []
+            for propname in proplist:
+                prop = runt.model.props.get(propname)
+                if prop is None:
+                    raise self.kids[0].addExcInfo(s_exc.NoSuchProp.init(propname))
+
+                pgenrs.append(self.pivogenr(runt, prop))
+
+            async def listpivot(node):
+                for pgenr in pgenrs:
+                    async for pivo in pgenr(node, strict=False):
+                        yield pivo
+
+            return(listpivot)
+
+        return self.pivogenr(runt, prop)
+
+    async def run(self, runt, genr):
+
+        pgenr = None
+        warned = False
+
+        async for node, path in genr:
+
+            if pgenr is None or not self.kids[0].isconst:
+                name = await self.kids[0].compute(runt, None)
+                pgenr = self.buildgenr(runt, name)
+
+            if self.isjoin:
+                yield node, path
 
-            if not found:
-                mesg = f'No pivot found for {node.form.name} -> {destform.name}.'
-                raise self.addExcInfo(s_exc.NoSuchPivot(n1=node.form.name, n2=destform.name, mesg=mesg))
+            try:
+                async for pivo in pgenr(node):
+                    yield pivo, path.fork(pivo)
+            except (s_exc.BadTypeValu, s_exc.BadLiftValu) as e:
+                if not warned:
+                    logger.warning(f'Caught error during pivot: {e.items()}')
+                    warned = True
+                items = e.items()
+                mesg = items.pop('mesg', '')
+                mesg = ': '.join((f'{e.__class__.__qualname__} [{repr(node.ndef[1])}] during pivot', mesg))
+                await runt.snap.fire('warn', mesg=mesg, **items)
```

Comment:
We could probably use `snap.warn()` now, instead of firing an event directly.

Reply:
updated to use snap.warn